garage/tf/algos/reps.py

```diff
@@ -0,0 +1,442 @@
+        self.optimizer.optimize(policy_opt_input_values)
+        logger.log("Computing policy KL")
+        policy_kl = self.f_policy_kl(*policy_opt_input_values)
+        logger.log("Computing policy loss after")
+        loss_after = self.optimizer.loss(policy_opt_input_values)
+        logger.record_tabular("EtaBefore", eta_before)
+        logger.record_tabular("EtaAfter", self.param_eta)
+        logger.record_tabular("DualBefore", dual_before)
+        logger.record_tabular("DualAfter", dual_after)
+        logger.record_tabular("{}/LossBefore".format(self.policy.name),
+                              loss_before)
+        logger.record_tabular("{}/LossAfter".format(self.policy.name),
+                              loss_after)
+        logger.record_tabular("{}/dLoss".format(self.policy.name),
+                              loss_before - loss_after)
+        logger.record_tabular("{}/KLBefore".format(self.policy.name),
+                              policy_kl_before)
+        logger.record_tabular("{}/KL".format(self.policy.name), policy_kl)
+
+    def _build_inputs(self):
+        """Decalre graph inputs variables."""
+        observation_space = self.policy.observation_space
+        action_space = self.policy.action_space
+        policy_dist = self.policy.distribution
+
+        with tf.name_scope("inputs"):
+            obs_var = observation_space.new_tensor_variable(
+                name="obs",
+                extra_dims=2)   # yapf: disable
+            action_var = action_space.new_tensor_variable(
+                name="action",
+                extra_dims=2)   # yapf: disable
+            reward_var = tensor_utils.new_tensor(
+                name="reward",
+                ndim=2,
+                dtype=tf.float32)   # yapf: disable
+            valid_var = tensor_utils.new_tensor(
+                name="valid",
+                ndim=2,
+                dtype=tf.float32)   # yapf: disable
+            feat_diff = tensor_utils.new_tensor(
+                name="feat_diff",
+                ndim=2,
+                dtype=tf.float32)   # yapf: disable
+            param_v = tensor_utils.new_tensor(
+                name="param_v",
+                ndim=1,
+                dtype=tf.float32)   # yapf: disable
+            param_eta = tensor_utils.new_tensor(
+                name="param_eta",
+                ndim=0,
+                dtype=tf.float32)   # yapf: disable
+            policy_state_info_vars = {
+                k: tf.placeholder(
+                    tf.float32,
+                    shape=[None] * 2 + list(shape),
+                    name=k)
+                for k, shape in self.policy.state_info_specs
+            }   # yapf: disable
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k]
+                for k in self.policy.state_info_keys
+            ]   # yapf: disable
+
+            policy_old_dist_info_vars = {
+                k: tf.placeholder(
```

Comment:
We could call [new_tensor](https://github.com/rlworkgroup/garage/blob/1c4eebe59e4609be1a890c968c293f2d592ed3ad/garage/tf/misc/tensor_utils.py#L72) here.

Reply:
em,,, we need to set `shape=[None] * 2 + list(shape)` here, but `new_tensor` returns a shape of `[None] * ndim`.

Reply:
oh, you're right, my bad